src/highdicom/seg/sop.py

```diff
@@ -2303,49 +2444,59 @@ def get_pixels_by_source_instance(
                 'uniquely identify frames of the segmentation image.'
             )
 
-        # Build the segmentation frame matrix
+        # Get a list of numerical indices for requested SOPInstanceUIDS
+        frame_indices = np.zeros(len(source_sop_instance_uids), np.uint32)
         for r, sop_uid in enumerate(source_sop_instance_uids):
             # Check whether this source UID exists in the LUT
             try:
                 src_uid_ind = self._get_src_uid_index(sop_uid)
             except KeyError as e:
                 if assert_missing_frames_are_empty:
                     continue
-                else:
-                    msg = (
-                        f'SOP Instance UID {sop_uid} does not match any '
-                        'referenced source frames. To return an empty '
-                        'segmentation mask in this situation, use the '
-                        "'assert_missing_frames_are_empty' parameter."
-                    )
-                    raise KeyError(msg) from e
+                msg = (
+                    f'SOP Instance UID {sop_uid} does not match any '
+                    'referenced source frames. To return an empty '
+                    'segmentation mask in this situation, use the '
+                    "'assert_missing_frames_are_empty' parameter."
+                )
+                raise KeyError(msg) from e
+            frame_indices[r] = src_uid_ind
+
+        # Loop through each row of the LUT and determine whether the
+        # corresponding segmentation frame is required in the output, and if so,
+        # where
+        for seg_frm, lut_row in enumerate(lut):
+            # Source frame index and segment number for this LUT row
+            lut_src, lut_segno = lut_row.ravel()
+
+            # Find indices (if any) of these value in the lists of required
+            # output source frames and segments
+            out_frm_inds = np.where(frame_indices == lut_src)[0]
+            out_seg_inds = np.where(segment_numbers == lut_segno)[0]
```

Comment:
I need to take a closer look, but at first glance I am wondering whether this is the most performant approach..

Reply:
See https://github.com/herrmannlab/highdicom/pull/208#issuecomment-1358495069